library/lua/gui/widgets/text_area.lua

```diff
@@ -0,0 +1,166 @@
+-- Multiline text area control
+
+local Panel = require('gui.widgets.containers.panel')
+local Scrollbar = require('gui.widgets.scrollbar')
+local TextAreaContent = require('gui.widgets.text_area.text_area_content')
+
+TextArea = defclass(TextArea, Panel)
+
+TextArea.ATTRS{
+    init_text = '',
+    init_cursor = DEFAULT_NIL,
+    text_pen = COLOR_LIGHTCYAN,
+    ignore_keys = {'STRING_A096'},
+    select_pen = COLOR_CYAN,
+    on_text_change = DEFAULT_NIL,
+    on_cursor_change = DEFAULT_NIL,
+    one_line_mode = false,
+    debug = false
+}
+
+function TextArea:init()
+    self.render_start_line_y = 1
+
+    self:addviews{
+        TextAreaContent{
+            view_id='text_area',
+            frame={l=0,r=3,t=0},
+            text=self.init_text,
+
+            text_pen=self.text_pen,
+            ignore_keys=self.ignore_keys,
+            select_pen=self.select_pen,
+            debug=self.debug,
+            one_line_mode=self.one_line_mode,
+
+            on_text_change=function (val)
+                self:updateLayout()
+                if self.on_text_change then
+                    self.on_text_change(val)
+                end
+            end,
+            on_cursor_change=self:callback('onCursorChange')
+        },
+        Scrollbar{
+            view_id='scrollbar',
+            frame={r=0,t=1},
+            on_scroll=self:callback('onScrollbar'),
+            visible=not self.one_line_mode
+        }
+    }
+    self:setFocus(true)
+end
+
+function TextArea:getText()
+    return self.subviews.text_area.text
+end
+
```

Comment:
Could we expose a `setText` method as well?

Reply:
done

Reply:
Could we make it so that if you call `setText(text)` and then press CTRL-Z, it will revert to the text before the `setText` call?

Reply:
I have mixed feelings about it, can you present a usecase where it would be useful?

Reply:
The situation where this came up for me is with this UI in a script I've been working on:

![image](https://github.com/user-attachments/assets/d5c18807-7053-45d3-98b1-1faa8cd51e03)

I'm using this to be able to quickly add notes to individual units. Once I no longer need the note for the unit, I'll clear it out with ALT+C. I know I could CTRL+A/backspace to clear all but I also wanted to provide a shortcut to clear the notes with a single hotkey press (I had this before while using an EditField here and still find it handy).

If a user ever did ALT+C by accident here, ideally they could CTRL+Z it.

Reply:
ok. I would additionally expose `clearHistory` function for cases where it would be a required behaviour.

Reply:
done